src/mq/gcloud-pubsub-push/mq.ts

```diff
@@ -0,0 +1,362 @@
+     * @param topic {string} Full name of topic to publish messages to
+     */
+    constructor(logger: Logger, pubSubClient: PubSub, topic: string) {
+        this.logger = logger;
+        this.pubSubClient = pubSubClient;
+        this.topic = topic;
+    }
+
+    /**
+     * Indicates whether the message queue is listening for messages or not
+     */
+    get isListening(): boolean {
+        return this.messageHandler !== undefined;
+    }
+
+    /**
+     * Enqueues a message
+     *
+     * @param message {any} Message to enqueue
+     * @param options {MessageQueueEnqueueOptions} Options for the enqueue operation
+     */
+    async enqueue(
+        message: any,
+        options?: MessageQueueEnqueueOptions,
+    ): Promise<void> {
+        const delay = options?.delay?.total('millisecond');
+
+        // If the message has a delay, do not enqueue it - This is likely a retry
+        // attempt by Fedify, but we do not want to retry the message as we want
+        // to use GCloud Pub/Sub's built in retry mechanism
+        if (delay !== undefined) {
+            this.logger.info(
+                `Not enqueuing message [FedifyID: ${message.id}] due to delay being set: ${delay}`,
+                { fedifyId: message.id },
+            );
+
+            return;
+        }
+
+        this.logger.info(`Enqueuing message [FedifyID: ${message.id}]`, {
+            fedifyId: message.id,
+        });
+
+        try {
+            const messageId = await this.pubSubClient
+                .topic(this.topic)
+                .publishMessage({
+                    json: message,
+                    attributes: {
+                        fedifyId: message.id,
+                    },
+                });
+
+            this.logger.info(
+                `Message [FedifyID: ${message.id}] was enqueued [PubSubID: ${messageId}]`,
+                { fedifyId: message.id, pubSubId: messageId },
+            );
+        } catch (error) {
+            this.logger.error(
+                `Failed to enqueue message [FedifyID: ${message.id}]: ${error}`,
+                { fedifyId: message.id, error },
+            );
+
+            this.errorListener?.(error as Error);
+        }
+    }
+
+    /**
+     * Starts the message queue
+     *
+     * @param handler {function} Message handler
+     * @param options {MessageQueueListenOptions} Options for the listen operation
+     */
+    async listen(
+        handler: (message: any) => Promise<void> | void,
+        options: MessageQueueListenOptions = {},
+    ): Promise<void> {
+        this.messageHandler = handler;
+
+        return await new Promise((resolve) => {
+            options.signal?.addEventListener('abort', () => {
+                resolve();
+            });
+        });
+    }
+
+    /**
+     * Handles a message
+     *
+     * @param message {Message} Message to handle
+     */
+    async handleMessage(message: Message): Promise<void> {
+        if (this.messageHandler === undefined) {
```

Comment:
Would be good to use `this.isListening` here

Reply:
Yeh I did try this, but further down where we use `this.messageHandler` it was giving me a type error about `this.messageHandler` being potentially undefined (which it wouldn't be as the `this.isListening` check would fail in that case). Any ideas how to get around this?